### PR TITLE
[Fix] [ttnn/moreh-defects]: adam exponent units, sgd tile count, softmin max-subtraction

### PR DESCRIPTION
**Pull request by ign-sudha**
## Summary

`Fixes: #51278`

Fixes the five remaining defects in that issue — items `1`, `2`, `11`, `12` and `13` — across `moreh_adam`, `moreh_sgd` and `moreh_softmax`.

**Items already on `main`**

The other eight items were fixed earlier, so this PR closes the issue.

| Item | Component | Fixed by |
| --- | --- | --- |
| `3` | `moreh_arange` row-major writer overrun | `ecd6b8f4202` (#56203) |
| `4` | `moreh_getitem` ROW_MAJOR `W`-index guard | `912d4b6c75d` (#52160) |
| `5` | `moreh_group_norm_backward` `gamma_grad` group index | `fce72e8060b` (#56202) |
| `6`, `7` | `moreh_nll_loss_step2` 2d weight guard, 3d loop bound | `9d2c9aa4b5a` (#56200) |
| `8`, `9`, `10` | `moreh_nll_loss_unreduced_backward` scratch CBs | `047fecfec7f` (#53534), renamed to DFBs by `54188ff85d0` (#54147) |

## Problem Description

**`moreh_adam` (items 1, 2)**

- `power_tile()` takes `param0` as IEEE-754 float bits, but the kernel passed the raw integer `step`. As float bits, `1` is a denormal (`≈1.4e-45`), so `pow(beta, step)` always evaluated to `1.0`, both bias corrections became `1/(1 - 1.0)`, and `param_out` came out `NaN`. `exp_avg` and `exp_avg_sq` were unaffected.
- Origin: `2fc9b5a4c55` (#33182) changed the `power_tile` contract and added `power_iterative_tile`, updating the callers in `moreh_common.hpp` but not this kernel.
- Separately, in the `sqrt` block `pack_tile_with_dt()` ran before `tile_regs_commit()` / `tile_regs_wait()`, so the packer read `DST` without the `MATH_PACK` handshake and could pick up a stale value from an earlier block. Every other block in the file packs after the handshake.

**`moreh_sgd` (item 11)**

- `H`, `W`, `Ht` and `Wt` came from `logical_shape()` while `num` divides the padded `physical_volume()`. For a shape whose `H` or `W` is not a multiple of `32` the integer division drops the last partial tile, so part of `param_out` is never written — and `[1, 1, 30, 32]` gives `Ht == 0`, i.e. no work at all. Sibling moreh factories use `padded_shape()` for tile math.

**`moreh_softmax` large-H / large-W (items 12, 13)**

- On the `SOFTMIN` path the last tile exponentiated the raw input instead of `(x - max)`, while every other tile and the whole `SOFTMAX` branch subtract the max first. `rexp_tile_and_mask_tile_to_cb()` does no max subtraction of its own, so the denominator mixed two scalings and each row came out scaled by a different wrong factor.

## What's Changed

**`moreh_adam.cpp`**

- Convert `step` to `float` and pass its bit pattern to both `power_tile()` calls, using the `union { float f; uint32_t u; }` idiom already in `reader_moreh_adam.cpp`. Host-side runtime args are untouched.
- Kept `power_tile` rather than switching to `power_iterative_tile`: the latter is an iterative multiply loop whose cost grows with the exponent, and Adam's `step` increases every optimizer iteration.
- Moved `pack_tile_with_dt()` to after `tile_regs_commit()` / `tile_regs_wait()`, matching the sibling blocks.

**`moreh_sgd_program_factory.cpp`**

- `param_in.logical_shape()` → `param_in.padded_shape()`. No masking needed: the op is elementwise, so the padded tile count simply covers the whole buffer. `compute_output_specs` still uses `logical_shape()`, so the output shape is unchanged.

**`moreh_softmax_h_large.cpp`, `moreh_softmax_w_large.cpp`**

- Hoisted `sub_tiles_bcast_rows_to_cb()` (H variant) and `sub_tiles_bcast_cols_to_cb()` (W variant) out of the last-tile branch, so every tile including the last is max-subtracted before exponentiating. This also removes the same call that was duplicated across all four branches in each file.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/moreh-fixes-51278)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/moreh-fixes-51278)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/moreh-fixes-51278)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/moreh-fixes-51278)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/moreh-fixes-51278)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/moreh-fixes-51278)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/moreh_adam.cpp [hunk 1]
```diff
@@ -21,6 +21,11 @@
 
 void kernel_main() {
     uint32_t step = get_arg_val<uint32_t>(0);
+    union {
```
Comment (iwroszTT) [inline ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/moreh_adam.cpp]: Compare this union based format conversion with the mechanism used in other compute kernels.

Comment (ign-yaswanth) [reply ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/moreh_adam.cpp]: The union was removed. Other moreh kernels raise an integer exponent with `power_iterative_tile`, so both beta1 and beta2 now call `power_iterative_tile(dst0, step)` and pass step through as an integer. A bit-cast into power_tile would still be the wrong contract here: that API reads the exponent as `IEEE-754` float bits and uses an approximate log/exp pow, which did not match the PyTorch update.

File: ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/moreh_adam.cpp [hunk 2]
```diff
@@ -146,8 +146,8 @@ void kernel_main() {
         tile_regs_acquire();
         copy_tile_init_with_dt(dfb_scalar_args_obj);
         copy_tile(cb_scalar_args, beta2_tile, dst0);
-        power_tile_init();
-        power_tile(dst0, step);
+        power_iterative_tile_init();
+        power_iterative_tile(dst0, step);
```
[the remark this answers is not in the document]
Comment (ign-yaswanth) [reply ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/moreh_adam.cpp]: Not a precision issue, so no numbers to add. step is an integer exponent and power_tile takes float bits via an approximate log/exp pow - wrong contract regardless of how close the result gets. Description updated to say that instead of "didn't match PyTorch in testing."

File: ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/moreh_adam.cpp [hunk 3]
```diff
@@ -220,10 +220,10 @@ void kernel_main() {
 #endif
         sqrt_tile_init();
         sqrt_tile(dst0);
-        pack_tile_with_dt(dst0, dfb_tmp1_obj);
         tile_regs_commit();
 
         tile_regs_wait();
+        pack_tile_with_dt(dst0, dfb_tmp1_obj);
```
[the remark this answers is not in the document]
Comment (ign-yaswanth) [reply ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/moreh_adam.cpp]: Checked this properly rather than adding a new loop. `test_moreh_adam` already runs this block up to 256 tiles per case across 128 parametrizations, and `param_out` is asserted on every one now. A repeat-N-times test wouldn't add coverage beyond that.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_program_factory.cpp [hunk 4]
```diff
@@ -43,7 +43,7 @@ ProgramDescriptor MorehSgdOperation::create_descriptor(
 
     auto compute_kernel_config = operation_attributes.compute_kernel_config;
 
-    auto shape = param_in.logical_shape();
+    auto shape = param_in.padded_shape();
```
[the remark this answers is not in the document]
Comment (ign-yaswanth) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_program_factory.cpp]: Checked - `momentum_buffer_out` is written in the same per-tile loop as `param_out` (writer_moreh_sgd.cpp), same tile count. One fix covers both.

File: ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/moreh_softmax_h_large.cpp [hunk 5]
```diff
@@ -62,10 +62,9 @@ void kernel_main() {
 
         for (std::uint32_t h = 0; h < Ht; h += onetile) {
             // compute exp(x - max(x))
```
[the remark this answers is not in the document]
Comment (ign-yaswanth) [reply ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/moreh_softmax_h_large.cpp]: Agreed it's not ideal, but leaving it - the comment labels the loop body as a whole, matching how the rest of the kernel is commented. Not worth a diff for this.

File: tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_adam.py [hunk 6]
```diff
@@ -66,8 +68,15 @@ def forward(self, x):
     cpu_grad = model.weight.grad.clone()
     dev_grad = create_tt_tensor(cpu_grad, device, dtype=dtype)
 
-    for _ in range(step):
-        optimizer.step()
+    # Device does one update at `step` from zero state; do the same on CPU.
+    if step > 1:
+        state = optimizer.state[model.weight]
+        state["step"] = torch.tensor(float(step - 1))
```
[the remark this answers is not in the document]
Comment (ign-yaswanth) [reply tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_adam.py]: Fixed - added a one-line comment explaining why: the kernel raises beta to step in one shot, it doesn't replay every step before it.

**Review by bbradelTT — APPROVED** [review]
